File: app/src/main/java/com/banew/cw2025_client/ui/login/LoginViewModel.java
package com.banew.cw2025_client.ui.login;

import android.util.Patterns;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.banew.cw2025_backend_common.dto.users.UserTokenFormResult;
import com.banew.cw2025_client.R;
import com.banew.cw2025_client.data.DataSource;
import com.banew.cw2025_client.data.Result;

public class LoginViewModel extends ViewModel {

    private final MutableLiveData<LoginFormState> loginFormState = new MutableLiveData<>();
    private final MutableLiveData<Result<UserTokenFormResult>> loginResult = new MutableLiveData<>();
    private final DataSource dataSource;

    LoginViewModel(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    LiveData<LoginFormState> getLoginFormState() {
        return loginFormState;
    }

    LiveData<Result<UserTokenFormResult>> getLoginResult() {
        return loginResult;
    }

    public void login(String username, String password) {
        // can be launched in a separate asynchronous job
        //dataSource.login(username, password).thenAccept(loginResult::setValue);
    }

    public void loginDataChanged(String username, String password) {
        if (!isUserNameValid(username)) {
            loginFormState.setValue(new LoginFormState(R.string.invalid_username, null));
        } else if (!isPasswordValid(password)) {
            loginFormState.setValue(new LoginFormState(null, R.string.invalid_password));
        } else {
            loginFormState.setValue(new LoginFormState(true));
        }
    }

    // A placeholder username validation check
    private boolean isUserNameValid(String username) {
        if (username == null) {
            return false;
        }
        return Patterns.EMAIL_ADDRESS.matcher(username).matches();
    }

    // A placeholder password validation check
    private boolean isPasswordValid(String password) {
        return password != null && password.trim().length() > 8;
    }
}

File: app/src/main/java/com/banew/cw2025_client/ui/start/LoginViewModel.kt
package com.banew.cw2025_client.ui.start

import android.util.Patterns
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.banew.cw2025_backend_common.dto.users.UserTokenFormResult
import com.banew.cw2025_client.GlobalApplication
import com.banew.cw2025_client.data.DataSource
import com.banew.cw2025_client.data.Result
import kotlinx.coroutines.launch

class LoginViewModel : ViewModel() {

    private val dataSource: DataSource? = GlobalApplication.getInstance()?.dataSource

    var email by mutableStateOf("")

    var password by mutableStateOf("")

    var username by mutableStateOf("")

    var photoSrc by mutableStateOf("")

    var isLoading by mutableStateOf(false)
        private set

    var loginResult by mutableStateOf<Result<UserTokenFormResult>?>(null)
        private set

    fun login() {
        viewModelScope.launch {
            isLoading = true
            loginResult = dataSource?.login(email, password)
            isLoading = false
        }
    }

    fun register() {
        viewModelScope.launch {
            isLoading = true
            loginResult = dataSource?.register(email, username, photoSrc, password)
            isLoading = false
        }
    }

    // A placeholder username validation check
    private fun isUserNameValid(username: String?): Boolean {
        if (username == null) {
            return false
        }
        return Patterns.EMAIL_ADDRESS.matcher(username).matches()
    }

    // A placeholder password validation check
    private fun isPasswordValid(password: String?): Boolean {
        return password != null && password.trim { it <= ' ' }.length > 8
    }
}